### DISCUSSION TITLE
BN256 G2 operations inside the circuit

### DISCUSSION THREAD

**Discussion by mharishankar** [Consensys/gnark#378] [Q&A]
Hi,
We have implemented BLS public key additions and BLS signature verification with BN256 in our solidity smart-contract - Public keys are in G2 and hence public key aggregation operations are G2 additions, while signatures are in G1. 

We are attempting to move the public key aggregations inside the snark as the BN256 G2 operations are gas expensive (https://github.com/musalbas/solidity-BN256G2) . However, I don't see support for this in the gnark repo, where only the twistededwards BN254 impl is available. Can you recommend a way to do this?

Thanks

**Comment by gbotrel** [comment]
Hello -- are you looking at doing BN256 Fp2 operations inside a BN256 Snark? If so, the field would be different, so unless I missed something in your goal, you would have to implement Fp2 arithmetic using the std/math/emulated package (develop branch), then point arithmetic.

**Reply by mharishankar** [reply]
Hey, thanks for the response. Yes I'm trying to do BN256 Fp2 operations inside the snark, correct. Can you elaborate a bit on how I can go about using the emulated/ package to add Fp2 arithmetic to the repo?

**Reply by mharishankar** [reply]
Hi, I have spent more time understanding this repo. I believe I should 1) I extend std/math/emulated/Params.go for BN256Fp2,  2) implement relevant arithmetic under std/algebra/fields_bn256 and std/algebra/sw_bn256. Is this correct?

**Reply by gbotrel** [reply]
quite close;

1. params.go -> BN254Fp  is enough to implement the extension. 
2. you can checkout this package for a reference: https://github.com/ConsenSys/gnark-crypto/tree/master/ecc/bn254/internal/fptower 
3. ```
type E2 struct {
A0, A1 emulated.Element[emulated.BN254Fp]
}
```
4. then yes, the missing piece: G2 arithmetic `std/algebra/sw_bn256` .

Note that the emulated arithmetic package is under development (need more optimizations) and performance (i.e number of constraints) will be sub-optimal for couple of weeks.

**Reply by mharishankar** [reply]
Hi Gautam,
Thanks for the help. I have implemented a very minimal set of changes to support just G2 negation in BN254 [here](https://github.com/nume-crypto/gnark/commit/33986b4d5c40eeb8e43a823076ef06ac8a11acb0) (branch **test**) . 

I added the relevant G2/E2 structs and Negation operations in fields_bn254 and sw_bn254 and extended field_test.go with the testcase (TestNegG2BN254).  However, this very simple testcase fails and the only thing I can think of is perhaps the emulated arithmetic in element.go is incorrect somewhere? Can you please take a look at this brief commit  and help?

One thing to note in your previous response - you suggested:
```
type E2 struct {
      A0, A1 emulated.Element[emulated.BN254Fp]
}
```
But I dont think  emulated.Element[emulated.BN254Fp] is required, based on the reference impl of fields_bls12-377/ and sw_12-377/  and the sample invocation of the 12-377 pairing operation in BN254 snark shown in TestPairingBLS377 in emulated/field_test.go

**Reply by gbotrel** [reply]
will have a look in couple of days -- 
But note that the `emulated` package in its tests may use `test.WithApiWrapper` or artifacts to compile the whole circuit as emulated (ie consider the witness input as emulated elements). I believe this was done for convenience reason (automatically testing all internal test circuits), but may not make a lot of sense in the future. 

The final API for emulated arithmetic will likely be revisited in the coming weeks when we iron out the latest perf detail.

**Reply by mharishankar** [reply]
Yup understood. Will wait for your thoughts on the my commit. Thanks again.

**Reply by mharishankar** [reply]
hey @gbotrel  did you have a chance to look at that commit? Can you weigh in on why it doesnt work / if im using the emulated arithmetic incorrectly?

**Reply by ivokub** [reply]
@mharishankar - I don't know if it is too late now, but I have implemented BN254 pairing in-circuit using field emulation. It may have relevant operations you are interested in. See https://github.com/ConsenSys/gnark/pull/411

**Reply by yelhousni** [reply]
The BN254 Fp2 arithmetic inside a SNARK is now here https://github.com/ConsenSys/gnark/blob/develop/std/algebra/emulated/fields_bn254/e2.go

You can add the G2 arithmetic here: https://github.com/ConsenSys/gnark/blob/develop/std/algebra/emulated/sw_bn254/g2.go

**Reply by ivokub** [reply]
Some G2 arithmetic is in this PR: https://github.com/ConsenSys/gnark/pull/629
It is still incomplete but you can get the gist.

**Reply by ivokub** [reply]
Tracking issue in https://github.com/ConsenSys/gnark/issues/647
